lib/wasi/src/runtime/module_cache/filesystem.rs

```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    pub fn cache_dir(&self) -> &Path {
+        &self.cache_dir
+    }
+
+    fn path(&self, key: ModuleHash, deterministic_id: &str) -> PathBuf {
+        let artifact_version = wasmer_types::MetadataHeader::CURRENT_VERSION;
+        self.cache_dir
+            .join(format!("{deterministic_id}-v{artifact_version}"))
+            .join(key.to_string())
+            .with_extension("bin")
+    }
+}
+
+#[async_trait::async_trait]
+impl ModuleCache for FileSystemCache {
+    async fn load(&self, key: ModuleHash, engine: &Engine) -> Result<Module, CacheError> {
+        let path = self.path(key, engine.deterministic_id());
+
+        // FIXME: This will all block the thread at the moment. Ideally,
+        // deserializing and uncompressing would happen on a thread pool in the
+        // background.
+        // https://github.com/wasmerio/wasmer/issues/3851
+
+        let uncompressed = read_compressed(&path)?;
+
+        match Module::deserialize_checked(&engine, &uncompressed) {
+            Ok(m) => Ok(m),
+            Err(e) => {
+                tracing::debug!(
+                    %key,
+                    path=%path.display(),
+                    error=&e as &dyn std::error::Error,
+                    "Deleting the cache file because the artifact couldn't be deserialized",
+                );
+
+                if let Err(e) = std::fs::remove_file(&path) {
+                    tracing::warn!(
+                        %key,
+                        path=%path.display(),
+                        error=&e as &dyn std::error::Error,
+                        "Unable to remove the corrupted cache file",
+                    );
+                }
+
+                Err(CacheError::Deserialize(e))
+            }
+        }
+    }
+
+    async fn save(
+        &self,
+        key: ModuleHash,
+        engine: &Engine,
+        module: &Module,
+    ) -> Result<(), CacheError> {
+        let path = self.path(key, engine.deterministic_id());
+
+        // FIXME: This will all block the thread at the moment. Ideally,
+        // serializing and compressing would happen on a thread pool in the
+        // background.
+        // https://github.com/wasmerio/wasmer/issues/3851
+
+        if let Some(parent) = path.parent() {
+            if let Err(e) = std::fs::create_dir_all(parent) {
+                tracing::warn!(
+                    dir=%parent.display(),
+                    error=&e as &dyn std::error::Error,
+                    "Unable to create the cache directory",
+                );
+            }
+        }
+
+        // Note: We'll first save to a temporary file in the same folder, then
+        // rename it when we are done serializing. Try to use a unique extension
+        // just in case we have concurrent saves of the same module.
+        static UNIQUE_ID: AtomicU64 = AtomicU64::new(0);
+        let extension = format!("tmp{}", UNIQUE_ID.fetch_add(1, Ordering::Relaxed));
```

Comment:
Multiple Wasmer processes could trample over each other with this. 

I'd use a `Uuid::new_v4()` or a nanosecond timestamp.

Reply:
I've switched to the `tempfile` crate. That should let us write to and persist temporary files properly without worrying about concurrent processes trampling each other.